bitcoin/src/witness_program.rs

```diff
@@ -20,29 +20,54 @@
 //! If the version byte is 0, but the witness program is neither 20 nor 32 bytes, the script must fail.
 //!
 
-use hex;
-use std::fmt;
 use std::str::FromStr;
 
+use wagyu_model::AddressError;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct WitnessProgram {
     /// The version byte
     pub version: u8,
     /// The witness program bytes
     pub program: Vec<u8>,
 }
 
+#[derive(Debug, Fail, PartialEq, Eq)]
+pub enum WitnessProgramError {
+    #[fail(display = "expected program with length at least 2, at most 40, actual program was length {}", _0)]
+    InvalidProgramLength(usize),
+
+    #[fail(display = "invalid program length {} for script version {}", _0, _1)]
+    InvalidProgramLengthForVersion(usize, u8),
+
+    #[fail(display = "expected version no greater than 16")]
+    InvalidVersion(u8),
+
+    #[fail(display = "expected program length {} to equal length {} as specified by input", _0, _1)]
+    MismatchedProgramLength(usize, usize),
+
+    #[fail(display = "error decoding program from hex string")]
+    ProgramDecodingError,
+}
+
+impl From<WitnessProgramError> for AddressError {
```

Comment:
Why is this here? Remove

Reply:
https://github.com/ArgusHQ/wagyu/blob/d455e031ff1866ff76d255c311d69ab5f7c39e9c/bitcoin/src/address.rs#L220

The `?` on this line requires this

Reply:
This should be under `model/address.rs` as it's an `impl` of `AddressError`. You'll find all of its cousin `{Model}Error` conversions there as well.

Reply:
resolved offline